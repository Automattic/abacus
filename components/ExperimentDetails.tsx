import debugFactory from 'debug'
import React from 'react'

import AudiencePanel from '@/components/AudiencePanel'
import GeneralPanel from '@/components/GeneralPanel'
import { ExperimentFull, Segment } from '@/models'

const debug = debugFactory('abacus:components/ExperimentDetails.tsx')

function ExperimentDetails({ experiment, segments }: { experiment: ExperimentFull; segments: Segment[] }) {
  debug('ExperimentDetails#render')
  return (
    <div>
      <h2>Experiment details</h2>
      <GeneralPanel experiment={experiment} />
      <br />
      <AudiencePanel experiment={experiment} segments={segments} />
    </div>
  )
}

export default ExperimentDetails
