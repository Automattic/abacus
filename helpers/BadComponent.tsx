import React from 'react'

const obj = {}

export default function BadComponent() {
  // Need a component that will fail when React attempts to render but also need
  // something that TypeScript will compile. The type assertion allows TS to
  // compile successfully but because we essentially lied to TS, it will fail as
  // runtime.
  return <>{(obj as { nonExistent: { boom: 'shaka laka' } }).nonExistent.boom}</>
}
