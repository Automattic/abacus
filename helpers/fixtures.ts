/**
 * Data fixtures to use in tests.
 *
 * Functions in this file return new objects populated with dummy values, which may potentially be overridden in
 * functions that accept Partial<T> as an argument.
 */
import _ from 'lodash'

import {
  Analysis,
  AnalysisStrategy,
  AttributionWindowSeconds,
  ExperimentFull,
  MetricAssignment,
  MetricBare,
  MetricFull,
  MetricRevenueParams,
  Platform,
  RecommendationReason,
  RecommendationWarning,
  Segment,
  SegmentAssignment,
  SegmentType,
  Status,
  TransactionTypes,
  Variation,
} from '@/models'

function createAnalysis(fieldOverrides: Partial<Analysis>) {
  return new Analysis({
    metricAssignmentId: 123,
    analysisStrategy: AnalysisStrategy.IttPure,
    participantStats: {
      total: 1000,
      not_final: 100,
      variation_1: 600,
      variation_2: 400,
    },
    metricEstimates: {
      diff: { estimate: 0.0, bottom: -0.01, top: 0.01 },
      variation_1: { estimate: 0.12, bottom: 0, top: 10.0 },
      variation_2: { estimate: -0.12, bottom: -1.123, top: 1.0 },
    },
    recommendation: {
      endExperiment: true,
      chosenVariationId: 2,
      reason: RecommendationReason.CiInRope,
      warnings: [RecommendationWarning.ShortPeriod, RecommendationWarning.WideCi],
    },
    analysisDatetime: new Date(Date.UTC(2020, 4, 10)),
    ...fieldOverrides,
  })
}

function createAnalyses() {
  return [
    // Full set of "latest" analyses for the default metric assignment.
    createAnalysis({
      analysisStrategy: AnalysisStrategy.IttPure,
      participantStats: {
        total: 1000,
        not_final: 100,
        variation_1: 600,
        variation_2: 400,
      },
    }),
    createAnalysis({
      analysisStrategy: AnalysisStrategy.MittNoCrossovers,
      participantStats: {
        total: 900,
        not_final: 90,
        variation_1: 540,
        variation_2: 360,
      },
      recommendation: {
        endExperiment: false,
        chosenVariationId: null,
        reason: RecommendationReason.RopeInCi,
        warnings: [],
      },
    }),
    createAnalysis({
      analysisStrategy: AnalysisStrategy.MittNoSpammers,
      participantStats: {
        total: 850,
        not_final: 85,
        variation_1: 510,
        variation_2: 340,
      },
      recommendation: {
        endExperiment: true,
        chosenVariationId: null,
        reason: RecommendationReason.CiInRope,
        warnings: [],
      },
    }),
    createAnalysis({
      analysisStrategy: AnalysisStrategy.MittNoSpammersNoCrossovers,
      participantStats: {
        total: 800,
        not_final: 80,
        variation_1: 480,
        variation_2: 320,
      },
    }),
    createAnalysis({
      analysisStrategy: AnalysisStrategy.PpNaive,
      participantStats: {
        total: 700,
        not_final: 70,
        variation_1: 420,
        variation_2: 280,
      },
    }),

    // Another set of analyses for the default metric assignment with an earlier date.
    createAnalysis({
      analysisStrategy: AnalysisStrategy.IttPure,
      participantStats: {
        total: 100,
        not_final: 10,
        variation_1: 60,
        variation_2: 40,
      },
      analysisDatetime: new Date(Date.UTC(2020, 4, 9)),
    }),
    createAnalysis({
      analysisStrategy: AnalysisStrategy.MittNoCrossovers,
      participantStats: {
        total: 90,
        not_final: 9,
        variation_1: 54,
        variation_2: 36,
      },
      recommendation: {
        endExperiment: false,
        chosenVariationId: null,
        reason: RecommendationReason.RopeInCi,
        warnings: [],
      },
      analysisDatetime: new Date(Date.UTC(2020, 4, 9)),
    }),
    createAnalysis({
      analysisStrategy: AnalysisStrategy.MittNoSpammers,
      participantStats: {
        total: 85,
        not_final: 8,
        variation_1: 51,
        variation_2: 34,
      },
      recommendation: {
        endExperiment: true,
        chosenVariationId: null,
        reason: RecommendationReason.CiInRope,
        warnings: [],
      },
      analysisDatetime: new Date(Date.UTC(2020, 4, 9)),
    }),
    createAnalysis({
      analysisStrategy: AnalysisStrategy.MittNoSpammersNoCrossovers,
      participantStats: {
        total: 80,
        not_final: 8,
        variation_1: 48,
        variation_2: 32,
      },
      analysisDatetime: new Date(Date.UTC(2020, 4, 9)),
    }),
    createAnalysis({
      analysisStrategy: AnalysisStrategy.PpNaive,
      participantStats: {
        total: 70,
        not_final: 7,
        variation_1: 42,
        variation_2: 28,
      },
      analysisDatetime: new Date(Date.UTC(2020, 4, 9)),
    }),

    // One example of a metric assignment with no data on one variation.
    createAnalysis({
      metricAssignmentId: 124,
      analysisStrategy: AnalysisStrategy.IttPure,
      participantStats: {
        total: 10,
        not_final: 10,
        variation_1: 10,
      },
      metricEstimates: null,
      recommendation: null,
    }),

    // Similar to the set of "latest" analyses for the default metric assignment, but with consistent recommendations.
    createAnalysis({
      metricAssignmentId: 126,
      analysisStrategy: AnalysisStrategy.IttPure,
      participantStats: {
        total: 2000,
        not_final: 200,
        variation_1: 1200,
        variation_2: 800,
      },
    }),
    createAnalysis({
      metricAssignmentId: 126,
      analysisStrategy: AnalysisStrategy.MittNoCrossovers,
      participantStats: {
        total: 1800,
        not_final: 180,
        variation_1: 1080,
        variation_2: 720,
      },
    }),
    createAnalysis({
      metricAssignmentId: 126,
      analysisStrategy: AnalysisStrategy.MittNoSpammers,
      participantStats: {
        total: 1700,
        not_final: 170,
        variation_1: 920,
        variation_2: 780,
      },
    }),
    createAnalysis({
      metricAssignmentId: 126,
      analysisStrategy: AnalysisStrategy.MittNoSpammersNoCrossovers,
      participantStats: {
        total: 1600,
        not_final: 160,
        variation_1: 960,
        variation_2: 640,
      },
    }),
    createAnalysis({
      metricAssignmentId: 126,
      analysisStrategy: AnalysisStrategy.PpNaive,
      participantStats: {
        total: 1400,
        not_final: 140,
        variation_1: 840,
        variation_2: 560,
      },
    }),
  ]
}

function createMetricAssignment(fieldOverrides: Partial<MetricAssignment>) {
  return new MetricAssignment({
    metricAssignmentId: 123,
    metricId: 1,
    experimentId: 1,
    attributionWindowSeconds: AttributionWindowSeconds.OneWeek,
    changeExpected: true,
    isPrimary: true,
    minDifference: 0.1,
    ...fieldOverrides,
  })
}

function createExperimentFull(fieldOverrides: Partial<ExperimentFull> = {}) {
  return new ExperimentFull({
    experimentId: 1,
    name: 'experiment_1',
    startDatetime: new Date(Date.UTC(2020, 5, 4)),
    endDatetime: new Date(Date.UTC(2020, 6, 4)),
    status: Status.Completed,
    platform: Platform.Calypso,
    ownerLogin: 'test_a11n',
    description: 'Experiment with things. Change stuff. Profit.',
    existingUsersAllowed: false,
    p2Url: 'https://wordpress.com/experiment_1',
    exposureEvents: null,
    variations: [
      new Variation({
        variationId: 2,
        name: 'test',
        isDefault: false,
        allocatedPercentage: 40,
      }),
      new Variation({
        variationId: 1,
        name: 'control',
        isDefault: true,
        allocatedPercentage: 60,
      }),
    ],
    metricAssignments: [
      createMetricAssignment({
        metricAssignmentId: 123,
        metricId: 1,
        attributionWindowSeconds: AttributionWindowSeconds.OneWeek,
        changeExpected: true,
        isPrimary: true,
        minDifference: 0.1,
      }),
      createMetricAssignment({
        metricAssignmentId: 124,
        metricId: 2,
        attributionWindowSeconds: AttributionWindowSeconds.FourWeeks,
        changeExpected: false,
        isPrimary: false,
        minDifference: 10.5,
      }),
      createMetricAssignment({
        metricAssignmentId: 125,
        metricId: 2,
        attributionWindowSeconds: AttributionWindowSeconds.OneHour,
        changeExpected: true,
        isPrimary: false,
        minDifference: 0.5,
      }),
      createMetricAssignment({
        metricAssignmentId: 126,
        metricId: 3,
        attributionWindowSeconds: AttributionWindowSeconds.SixHours,
        changeExpected: true,
        isPrimary: false,
        minDifference: 12,
      }),
    ],
    segmentAssignments: [],
    ...fieldOverrides,
  })
}

function createMetricBare(id: number) {
  return new MetricBare({
    metricId: id,
    name: `metric_${id}`,
    description: `This is metric ${id}`,
    parameterType: id % 2 === 0 ? 'revenue' : 'conversion',
  })
}

function createMetricBares(numMetrics = 3) {
  return _.range(1, numMetrics + 1).map(createMetricBare)
}

function createMetricFull(id: number) {
  // Note: It is hard to reuse createMetricBare here as it is boxed
  //       Currently we only unbox it into an ApiData format which is different from this
  const parameterType = id % 2 === 0 ? 'revenue' : 'conversion'
  const eventParams = [{ event: 'event_name', props: { has_blocks: 'true' } }]
  const revenueParams = new MetricRevenueParams({
    refundDays: id * 2,
    productSlugs: ['xx-bundles'],
    transactionTypes: [TransactionTypes.NewPurchase],
  })
  return new MetricFull({
    metricId: id,
    name: `metric_${id}`,
    description: `This is metric ${id}`,
    parameterType,
    higherIsBetter: id % 3 === 0,
    eventParams: parameterType === 'conversion' ? eventParams : null,
    revenueParams: parameterType === 'revenue' ? revenueParams : null,
  })
}

function createSegment(id: number) {
  return new Segment({
    segmentId: id,
    name: `segment_${id}`,
    type: id % 2 === 0 ? SegmentType.Country : SegmentType.Locale,
  })
}

/**
 * Creates an array of segments.
 */
function createSegments(numSegments: number) {
  return _.range(numSegments).map(createSegment)
}

function createSegmentAssignment(fieldOverrides: Partial<SegmentAssignment>) {
  return new SegmentAssignment({
    segmentAssignmentId: 123,
    experimentId: 1,
    segmentId: 1,
    isExcluded: false,
    ...fieldOverrides,
  })
}

const Fixtures = {
  createAnalyses,
  createExperimentFull,
  createMetricAssignment,
  createMetricBares,
  createMetricFull,
  createSegmentAssignment,
  createSegments,
}

export default Fixtures
