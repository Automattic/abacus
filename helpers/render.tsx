import { Queries, render as actualRender, RenderOptions } from '@testing-library/react'
import React from 'react'

import ThemeProvider from '@/styles/ThemeProvider'

const render: typeof actualRender = <Q extends Queries>(ui: React.ReactElement, options?: RenderOptions<Q>) =>
  actualRender((<ThemeProvider>{ui}</ThemeProvider>) as React.ReactElement, options) as ReturnType<typeof actualRender>

export default render
