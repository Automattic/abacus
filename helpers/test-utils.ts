import mediaQuery from 'css-mediaquery'

/**
 * Create a `matchMedia` function that will match a query based on the specified
 * width.
 *
 *     const initialJsDomWindowInnerWidth = window.innerWidth
 *     afterEach(() => {
 *       // Reset back to initial width for tests that don't explicitly set the width
 *       // themselves.
 *       window.matchMedia = createMatchMedia(initialJsDomWindowInnerWidth)
 *     })
 *
 *     test('...', () => {
 *       window.matchMedia = createMatchMedia(600)
 *       ...
 *     })
 *
 * @param width - The width of the window to simulate.
 */
function createMatchMedia(width: number) {
  return (query: string) => ({
    matches: mediaQuery.match(query, { width }),
    media: query,
    addEventListener: jest.fn(),
    addListener: jest.fn(),
    dispatchEvent: jest.fn(),
    onchange: jest.fn(),
    removeEventListener: jest.fn(),
    removeListener: jest.fn(),
  })
}

export { createMatchMedia }
