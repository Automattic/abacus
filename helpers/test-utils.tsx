import { Queries, render as actualRender, RenderOptions } from '@testing-library/react'
import mediaQuery from 'css-mediaquery'
import React from 'react'

import ThemeProvider from '@/styles/ThemeProvider'

/**
 * A wrapped unit-test react-renderer, useful for adding React Contexts globally.
 */
export const render: typeof actualRender = <Q extends Queries>(ui: React.ReactElement, options?: RenderOptions<Q>) =>
  actualRender((<ThemeProvider>{ui}</ThemeProvider>) as React.ReactElement, options) as ReturnType<typeof actualRender>

/**
 * Create a `matchMedia` function that will match a query based on the specified
 * width.
 *
 *     const initialJsDomWindowInnerWidth = window.innerWidth
 *     afterEach(() => {
 *       // Reset back to initial width for tests that don't explicitly set the width
 *       // themselves.
 *       window.matchMedia = createMatchMedia(initialJsDomWindowInnerWidth)
 *     })
 *
 *     test('...', () => {
 *       window.matchMedia = createMatchMedia(600)
 *       ...
 *     })
 *
 * @param width - The width of the window to simulate.
 */
export function createMatchMedia(width: number) {
  return (query: string) => ({
    matches: mediaQuery.match(query, { width }),
    media: query,
    addEventListener: jest.fn(),
    addListener: jest.fn(),
    dispatchEvent: jest.fn(),
    onchange: jest.fn(),
    removeEventListener: jest.fn(),
    removeListener: jest.fn(),
  })
}
