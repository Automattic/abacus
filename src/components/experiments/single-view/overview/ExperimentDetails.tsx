import Grid from '@material-ui/core/Grid'
import { useTheme } from '@material-ui/core/styles'
import useMediaQuery from '@material-ui/core/useMediaQuery'
import debugFactory from 'debug'
import React from 'react'

import AudiencePanel from 'src/components/experiments/single-view/overview/AudiencePanel'
import ConclusionsPanel from 'src/components/experiments/single-view/overview/ConclusionsPanel'
import GeneralPanel from 'src/components/experiments/single-view/overview/GeneralPanel'
import MetricAssignmentsPanel from 'src/components/experiments/single-view/overview/MetricAssignmentsPanel'
import { ExperimentFull, MetricBare, Segment, Status, TagBare } from 'src/lib/schemas'

const debug = debugFactory('abacus:components/ExperimentDetails.tsx')

/**
 * Renders the main details of an experiment.
 */
function ExperimentDetails({
  experiment,
  metrics,
  segments,
  tags,
  experimentReloadRef,
}: {
  experiment: ExperimentFull
  metrics: MetricBare[]
  segments: Segment[]
  tags: TagBare[]
  experimentReloadRef: React.MutableRefObject<() => void>
}): JSX.Element {
  debug('ExperimentDetails#render')
  const theme = useTheme()
  const isMdDown = useMediaQuery(theme.breakpoints.down('md'))

  return (
    <Grid container spacing={2}>
      <Grid item xs={12} lg={7}>
        <Grid container direction='column' spacing={2}>
          <Grid item>
            <GeneralPanel {...{ experiment, experimentReloadRef }} />
          </Grid>
          {isMdDown && (
            <Grid item>
              <AudiencePanel {...{ experiment, segments, tags }} />
            </Grid>
          )}
          <Grid item>
            <MetricAssignmentsPanel {...{ experiment, metrics, experimentReloadRef }} />
          </Grid>
          {(experiment.status === Status.Completed || experiment.status === Status.Disabled) && (
            <Grid item>
              <ConclusionsPanel {...{ experiment, experimentReloadRef }} />
            </Grid>
          )}
        </Grid>
      </Grid>
      {!isMdDown && (
        <Grid item lg={5}>
          <AudiencePanel {...{ experiment, segments, tags }} />
        </Grid>
      )}
    </Grid>
  )
}

export default ExperimentDetails
