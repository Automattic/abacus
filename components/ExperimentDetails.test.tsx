import { render } from '@testing-library/react'
import React from 'react'

import Fixtures from '@/helpers/fixtures'
import { createMatchMedia } from '@/helpers/test-utils'

import ExperimentDetails from './ExperimentDetails'

const initialJsDomWindowInnerWidth = window.innerWidth
afterEach(() => {
  window.matchMedia = createMatchMedia(initialJsDomWindowInnerWidth)
})

test('renders as expected at large width', () => {
  window.matchMedia = createMatchMedia(1600)

  const metrics = Fixtures.createMetricBares()
  const segments = Fixtures.createSegments(5)
  const experiment = Fixtures.createExperimentFull({
    segmentAssignments: [
      Fixtures.createSegmentAssignment({ segmentAssignmentId: 101, segmentId: 1 }),
      Fixtures.createSegmentAssignment({ segmentAssignmentId: 102, segmentId: 2, isExcluded: true }),
    ],
  })
  const { container } = render(<ExperimentDetails experiment={experiment} metrics={metrics} segments={segments} />)

  expect(container).toMatchSnapshot()
})

test('renders as expected at small width', () => {
  window.matchMedia = createMatchMedia(600)

  const metrics = Fixtures.createMetricBares()
  const segments = Fixtures.createSegments(5)
  const experiment = Fixtures.createExperimentFull({
    segmentAssignments: [
      Fixtures.createSegmentAssignment({ segmentAssignmentId: 101, segmentId: 1 }),
      Fixtures.createSegmentAssignment({ segmentAssignmentId: 102, segmentId: 2, isExcluded: true }),
    ],
  })
  const { container } = render(<ExperimentDetails experiment={experiment} metrics={metrics} segments={segments} />)

  expect(container).toMatchSnapshot()
})
