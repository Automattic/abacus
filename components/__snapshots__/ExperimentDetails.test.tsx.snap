// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders as expected 1`] = `
<div>
  <div>
    <h2>
      Experiment details
    </h2>
    <div
      class="MuiPaper-root MuiPaper-elevation1 MuiPaper-rounded"
    >
      <h3
        class="MuiTypography-root makeStyles-title-2 MuiTypography-h3 MuiTypography-colorTextPrimary"
      >
        General
      </h3>
      <table
        class="MuiTable-root"
      >
        <tbody
          class="MuiTableBody-root"
        >
          <tr
            class="MuiTableRow-root"
          >
            <th
              class="MuiTableCell-root MuiTableCell-head"
              role="cell"
              scope="row"
            >
              Description
            </th>
            <td
              class="MuiTableCell-root MuiTableCell-body"
            >
              Experiment with things. Change stuff. Profit.
            </td>
          </tr>
          <tr
            class="MuiTableRow-root"
          >
            <th
              class="MuiTableCell-root MuiTableCell-head"
              role="cell"
              scope="row"
            >
              P2 Link
            </th>
            <td
              class="MuiTableCell-root MuiTableCell-body"
            >
              <a
                href="https://wordpress.com/experiment_1"
                rel="noopener noreferrer"
                target="_blank"
              >
                https://wordpress.com/experiment_1
              </a>
            </td>
          </tr>
          <tr
            class="MuiTableRow-root"
          >
            <th
              class="MuiTableCell-root MuiTableCell-head"
              role="cell"
              scope="row"
            >
              Dates
            </th>
            <td
              class="MuiTableCell-root MuiTableCell-body"
            >
              <span
                class="makeStyles-root-3"
                title="03/06/2020, 20:00:00"
              >
                2020-06-04
              </span>
              <span
                class="makeStyles-to-1"
              >
                to
              </span>
              <span
                class="makeStyles-root-3"
                title="03/07/2020, 20:00:00"
              >
                2020-07-04
              </span>
            </td>
          </tr>
          <tr
            class="MuiTableRow-root"
          >
            <th
              class="MuiTableCell-root MuiTableCell-head"
              role="cell"
              scope="row"
            >
              Owner
            </th>
            <td
              class="MuiTableCell-root MuiTableCell-body"
            >
              test_a11n
            </td>
          </tr>
        </tbody>
      </table>
    </div>
    <br />
    <div
      class="MuiPaper-root MuiPaper-elevation1 MuiPaper-rounded"
    >
      <h3
        class="MuiTypography-root makeStyles-title-2 MuiTypography-h3 MuiTypography-colorTextPrimary"
      >
        Audience
      </h3>
      <table
        class="MuiTable-root"
      >
        <tbody
          class="MuiTableBody-root"
        >
          <tr
            class="MuiTableRow-root"
          >
            <th
              class="MuiTableCell-root MuiTableCell-head"
              role="cell"
              scope="row"
            >
              Platform
            </th>
            <td
              class="MuiTableCell-root MuiTableCell-body"
            >
              calypso
            </td>
          </tr>
          <tr
            class="MuiTableRow-root"
          >
            <th
              class="MuiTableCell-root MuiTableCell-head"
              role="cell"
              scope="row"
            >
              User Type
            </th>
            <td
              class="MuiTableCell-root MuiTableCell-body"
            >
              New users only
            </td>
          </tr>
          <tr
            class="MuiTableRow-root"
          >
            <th
              class="MuiTableCell-root MuiTableCell-head"
              role="cell"
              scope="row"
            >
              Variations
            </th>
            <td
              class="MuiTableCell-root MuiTableCell-body MuiTableCell-paddingNone"
            >
              <table
                class="MuiTable-root"
              >
                <thead
                  class="MuiTableHead-root"
                >
                  <tr
                    class="MuiTableRow-root MuiTableRow-head"
                  >
                    <th
                      class="MuiTableCell-root MuiTableCell-head"
                      role="columnheader"
                      scope="col"
                    >
                      Name
                    </th>
                    <th
                      class="MuiTableCell-root MuiTableCell-head"
                      role="columnheader"
                      scope="col"
                    >
                      Percent
                    </th>
                  </tr>
                </thead>
                <tbody
                  class="MuiTableBody-root"
                >
                  <tr
                    class="MuiTableRow-root"
                  >
                    <td
                      class="MuiTableCell-root MuiTableCell-body"
                    >
                      control
                      <div
                        class="MuiChip-root makeStyles-default-4"
                      >
                        <span
                          class="MuiChip-label"
                        >
                          Default
                        </span>
                      </div>
                    </td>
                    <td
                      class="MuiTableCell-root MuiTableCell-body"
                    >
                      60
                      %
                    </td>
                  </tr>
                  <tr
                    class="MuiTableRow-root"
                  >
                    <td
                      class="MuiTableCell-root MuiTableCell-body"
                    >
                      test
                    </td>
                    <td
                      class="MuiTableCell-root MuiTableCell-body"
                    >
                      40
                      %
                    </td>
                  </tr>
                </tbody>
              </table>
            </td>
          </tr>
          <tr
            class="MuiTableRow-root"
          >
            <th
              class="MuiTableCell-root MuiTableCell-head"
              role="cell"
              scope="row"
            >
              Segments
            </th>
            <td
              class="MuiTableCell-root MuiTableCell-body MuiTableCell-paddingNone"
            >
              <table
                class="MuiTable-root"
              >
                <thead
                  class="MuiTableHead-root"
                >
                  <tr
                    class="MuiTableRow-root MuiTableRow-head"
                  >
                    <th
                      class="MuiTableCell-root MuiTableCell-head"
                      role="columnheader"
                      scope="col"
                    >
                      Locales
                    </th>
                  </tr>
                </thead>
                <tbody
                  class="MuiTableBody-root"
                >
                  <tr
                    class="MuiTableRow-root"
                  >
                    <td
                      class="MuiTableCell-root MuiTableCell-body"
                    >
                      segment_1
                    </td>
                  </tr>
                </tbody>
              </table>
              <table
                class="MuiTable-root"
              >
                <thead
                  class="MuiTableHead-root"
                >
                  <tr
                    class="MuiTableRow-root MuiTableRow-head"
                  >
                    <th
                      class="MuiTableCell-root MuiTableCell-head"
                      role="columnheader"
                      scope="col"
                    >
                      Countries
                    </th>
                  </tr>
                </thead>
                <tbody
                  class="MuiTableBody-root"
                >
                  <tr
                    class="MuiTableRow-root"
                  >
                    <td
                      class="MuiTableCell-root MuiTableCell-body"
                    >
                      segment_2
                      <div
                        class="MuiChip-root makeStyles-excluded-5"
                      >
                        <span
                          class="MuiChip-label"
                        >
                          Excluded
                        </span>
                      </div>
                    </td>
                  </tr>
                </tbody>
              </table>
            </td>
          </tr>
        </tbody>
      </table>
    </div>
  </div>
</div>
`;
